README.md
# Derver

Tiny Development Server for your web-applications with livereload and watchers


<p align="center">
  <img src="https://github.com/AlexxNB/derver/raw/main/demo.gif">
</p>

## Features

* Very tiny (~8kb)
* Livereload
* Watch directories with callback
* CLI or JS API
* Supports Gzip or Brotli compression
* Cache control
* Supports SPA mode

## CLI Usage

```sh
derver [parameters] [serve_directory]
```

You may use `npx derver` instantly  or install globally by `npm install -g derver` and then use command `derver`.

To serve and watching current directory just run:

```sh
derver
```

Server will run on [http://localhost:7000]() and site will be reloaded each time you change the file in the served directory.

By default server is running on `localhost` and nobody can access on the website from the network. Bind server on the `0.0.0.0` interface to allow network connections:

```sh
derver --host=0.0.0.0 public
```

In this example, `public` is a directory where are files for serving. By default it is current directory.

### List of possible parameters

#### `--host`
Interface, where bind the server. Use `0.0.0.0` inside Docker container or when need network connections to your site.
*Default: localhost*
*Example: --host=localhost*

#### `--port`
Port, where bind the server. 
*Default: 7000*
*Example: --port=8080*

#### `--index`
Name of the root file of web directory. Webserver will lookup this file when no file specified in the requested URL. 
*Default: index.html*
*Example: --index=index.htm*

#### `--watch`
Specify the directories for watching files changes. Each time when files modified in these directories, website will be reloaded. You may use this parameter multiple times to set more than one directory for watching.
*Default: watch the served directory*
*Example: --watch=dist/public --watch=src*

#### `--no-watch`
Add this parameter when you want to disable any watching and livereloading. 
*Example: --no-watch*

#### `--compress`
Will return files compressed by gzip or brotli, if client supports it.
*Example: --compress*

#### `--cache`
Add `Cache-control` header to the response with `max-age` equal `31536000` (~1 year). You can specify number of seconds.
*Example: --cache*
*Example: --cache=3600*

#### `--scroll`
Restore scroll position on every reload. Number value is equal timeout before scroll restoration.
*Example: --scroll*
*Example: --scroll=100*

#### `--spa`
Enables SPA (Single-Page Application) mode. All requested pages will be responsed by index page in the application root, which is specified in `--index` parameter.
*Example: --spa*

#### `--production`
Run server in production mode(really not, use at own risk). It enables `--cache`, `--compress` and `--no-watch` parameters. Also host will set on `0.0.0.0` to handle connections from the network.

*Example: --production*

## Javascript API

You may use Derver in the your scripts to get more power.

Install Derver as local dependency:

```
npm install derver
```

Then use `derver` function from the `derver` package to start the server. 

```js
import {derver} from 'derver';

derver();
```

By default, server will be started on [http://localhost:7000]() and serve `public` directory in your workdir.

### Configuration

You may set configuration object as a parameter of the `derver` function. Below you find all possible options:

#### `dir` *string*|boolean
Directory which contains files for serving. If nothing set in `watch` option, it will be watching for changes also. When it is `false` - no files would be serving, only middlewares will work. 


*Default: public*


*Example: dir: 'public'*
*Example: dir: false*

---

#### `host` *string*
Interface, where bind the server. Use `0.0.0.0` inside docker or when need network connections to your site.


*Default: localhost*


*Example: host: 'localhost'*

---

#### `port` *number*
Port, where bind the server. 


*Default: 7000*


*Example: port: 8080*

---

#### `index`  *string*
Name of the root file of web directory. Webserver will lookup this file when no file specified in the requested URL. 

*Default: index.html*

*Example: index: 'index.htm'*

---

#### `compress` *boolean*
Will return files compressed by gzip or brotli, if client supports it.


*Default: false*


*Example: compress: true*

---

#### `cache` *boolean*|*number*
Add `Cache-control` header to the response with `max-age` equal `31536000` (~1 year). You can specify number of seconds.


*Default: false*


*Example: cache: true*


*Example: cache: 3600*

---

#### `spa` *boolean*
Enables SPA (Single-Page Application) mode. All requested pages will be responced by index page in the application root, which is specified in `index` option.


*Default: false*


*Example: spa: true*

---

#### `watch` *string*|*array of string*
Specify the directories for watching filechanges. Each time when files modified in theese directories, website will be reloaded and `onwatch` callback will be run. By default will be watched directory defined in `dir` option.


*Default: null*


*Example: watch: ['dist/public','src']*

---

#### `remote` *boolean*|*string*
Enables remote control listener. See [Remote control](#remote-control)


*Default: false*


*Example: remote: true*


*Example: remote: "my_dev_server"*

---

#### `parseJson` *boolean*
When incoming request sent with type `application/json` Derver will parse its body and put object in `request.body`.


*Default: true*


*Example: parseJson: true*

---

#### `preserveScroll` *boolean*|*number*

Restore scroll position on every reload. Number value is equal timeout before scroll restoration.

*Default: false*


*Example: preserveScroll: true*


*Example: preserveScroll: 100*

---

#### `banner` *boolean*
Show or not the banner in console when server starts.


*Default: true*


*Example: banner: false*

---

#### `log` *boolean*
Show or not file requests in console


*Default: true*


*Example: log: false*

---

#### `onwatch` *function*
This function will be called when any file changes in watched directories.


*Default: null*


*Example: onwatch: (liverload,watchitem)=>{if(watchitem == 'src') livereload.prevent()})*

---

### `onwatch`-callback arguments

*Callback signature: (livereload,watchitem,filename,eventname)*

#### `livereload`
It is object with following methods:

* `livereload.prevent()` - Will stop sheduled livereload action for this watch event.
* `livereload.reload()` - Run each time you want to reload page in the browser.
* `livereload.console(message)` - Send a `message` to the browser console.
* `livereload.error(message,[header])` - Show error modal on client.

---

#### `watchitem`
It is a string with directory name where were fired filechange event. It is same string as you specified in `watch` option(or in `dir` option, if `watch` not set).

---

#### `filename`
Full path of changed file (unstable)

---

#### `eventname`
What exactly happened with modified file. 


## Using Middlewares

You may use any common middleware(like Express middlewares) to add additional functionality for you server. `derver()` function returns the object with methods:
 - `sub` - run callback to register middlewares for specified subpath
 - `use` - run middleware for all HTTP methods
 - `get` - run middleware for GET method only
 - `post` - run middleware for POST method only
 - ...actually you can write any HTTP method here

 ```js
 derver()
  .use(middleware1)
  .get('/api',middleware2)
  .put('/clear',middleware3,middleware4)

 ```

 ### Pattern

 If first argument for these methods is a pattern of the URL, middleware will run only if request's URL is matched with its path. 

 The pattern may looks like `/foo` of `/foo/bar`. If no pattern provided, middleware will run on each request.

 Pattern also may have a parameters `/user/:name` and when URL will be `/user/bob` or `/user/alex` you can get the value(alex or bob) from `request.params.name` property.

  ```js
 derver()
  .use('/user/:name',middleware)

 ```

 ### Writing middleware function 

 The middleware functions gets three arguments - common Node's `request`,`response` objects and `next` function, which will run next middleware. If your middleware doesn't response on request, you must run `next()` or request never will be ended.

 Lets see an example for client and API middleware:

 ```js
  function myLogMiddleware(req,resp,next){
    console.log('Current URL is: ' + req.url);
    next();
  }

  function myHelloMiddleware(req,resp){
    resp.send('Hello, '+req.params.name);
  }

  derver()
    .use(myLogMiddleware)
    .get('/hello/:name',myHelloMiddleware)
 ```

### Additional data from incoming request extension

The `request` argument is Node's [http.IncomingMessage object](https://nodejs.org/api/http.html#http_class_http_incomingmessage). But it is expanded with few useful data:
* `path` - pathname of current URL 
* `search` - query params as a string
* `query` - query params as an object
* `host` - host from header(including port)
* `hostname` - host from header(without port)
* `port` - port from header

If request sent with type `application/json` you will get the parsed object from `request.body`. To avoid this, set `parseJson` option to `false`;

### Send JSON object in responce

Method `responce.send(message)` will send content of message with status code 200. If `message` is a simple object, Derver will automaticly stringify it and send to client with `Content-type: application/json` header.
 
### Nested middlewares 

In case you need to run middlewares which are situated under specified sub path use `.sub()` method.

```js

derver()
  .sub('/api',(app)=>
    // will run on  every request starting with '/api/...'
    app.use(myLogMiddleware);

    // will run when URL will be '/api/hello/bob'
    app.get('/hello/bob',myHelloMiddleware);

    app.sub('/users',(app)=>{
        // will run when URL will be '/api/users/add'
        app.post('/add',myUserAddMiddleware);
    })
  })

```

## Remote control

There is a way to perform some actions in currently opened browser windows. For example you want to reload page from external script.

```js
import {createRemote} from 'derver';

// Create remote object with parameters of running server
const remote = createRemote({host:'localhost',port:7000}); // there are defaults, may be dropped

// Also you can call `createRemote('my_dev_server)` with server ID specified in remote option.

// Reload page in all opened browser windows
remote.reload();

// Send some text in the browser console
remote.console('Hello!');

// Show modal with error message
remote.error('Error happened! Fix it as soon as possible!','Error header');
```

*Note: don't forget to enable `remote` option in Derver's configuration*


## How livereload works

When you changes file in the watching directory, server will send command to the client side to reload current page. It is musthave feature when you developing web-application and want to see changes immediately.

Livereload made with [Server Sent Events API](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events). It is perfect feature for one-way communication server with client based only on http protocol. It is why Derver is so tiny, no need to implement websocket communication as others known servers do.

Some JavaScript code for livereload will be added before `</body>` element inside each requested `html` file.

CHANGELOG.md
# Derver changelog

## unreleased

### Other

- Merge branch 'main' of github.com:AlexxNB/derver [`6f8d7d97`](https://github.com/AlexxNB/derver/commit/6f8d7d9750ee830e0e08bd2e84d6b0854beef340)

## 0.5.3 - 2021-11-11

### Other

- add node path [`1dcd969c`](https://github.com/AlexxNB/derver/commit/1dcd969c3594fddb38ef9033ffc0591ea661b44d)
- Merge branch 'main' of github.com:AlexxNB/derver [`04ddf9fd`](https://github.com/AlexxNB/derver/commit/04ddf9fd2570970f3ea7299ad194ed40c05541fe)

## 0.5.2 - 2021-11-06

### Features

- Preserve scroll position on reload [`483cc95a`](https://github.com/AlexxNB/derver/commit/483cc95ac5a9877976c104cd9ef79ba9eb6f3113)

### Other

- Merge branch 'main' of github.com:AlexxNB/derver [`e26d23c7`](https://github.com/AlexxNB/derver/commit/e26d23c79a01ae5cd2b08dace8a406bcd3be5633)

## 0.5.1 - 2021-11-01

### Bug Fixes

- 🐛 Add exports for main and module also [`efaccc30`](https://github.com/AlexxNB/derver/commit/efaccc307a062ae5394edebfe46f458771636c76)

### Other

- Merge branch 'main' of github.com:AlexxNB/derver [`6cb79cb4`](https://github.com/AlexxNB/derver/commit/6cb79cb444efa9aba36fce0fcfa20ddebde62040)

## 0.4.20 - 2021-10-30

### Other

- make rollup-plugin universal fo cjs and esm consumers [`48f130b6`](https://github.com/AlexxNB/derver/commit/48f130b6a6fbfcdbe7ad5b678afcb7905917f1a0)
- Merge branch 'main' of github.com:AlexxNB/derver into main [`3779af12`](https://github.com/AlexxNB/derver/commit/3779af128dd0a67bf42ec5926834ba2afbc04159)

## 0.4.19 - 2021-09-01

### Features

- Add options to prevent console trashing [`6883c207`](https://github.com/AlexxNB/derver/commit/6883c207d93c82638000a73f3696d23e5bb88a44)

## 0.4.18 - 2021-09-01

### Features

- Ability to run only middlewares when dir option is false [`1c3c5bc1`](https://github.com/AlexxNB/derver/commit/1c3c5bc1a813369fca2338b9398f01e8e9a48bfb)

### Other

- Merge branch 'main' of github.com:AlexxNB/derver into main [`082778e8`](https://github.com/AlexxNB/derver/commit/082778e8fc04c59de3eaca59dc6bd6ab3475f7bd)

## 0.4.17 - 2021-06-28

### Features

- Add JSON handling in middlewares [`f967e600`](https://github.com/AlexxNB/derver/commit/f967e600240397155148ae9a0ca23754ebb7e81c)

### Other

- Merge branch 'main' of github.com:AlexxNB/derver into main [`554a0eb8`](https://github.com/AlexxNB/derver/commit/554a0eb89f1c7407316dfb76016cb943a53de207)

## 0.4.16 - 2021-06-12

### Documentation

- 📝 Fix some mistakes [`cbc6d30e`](https://github.com/AlexxNB/derver/commit/cbc6d30eb5f5bc61190df6e8238a4c003ecda927)

### Other

- Merge branch 'main' of github.com:AlexxNB/derver into main [`0b76ccdf`](https://github.com/AlexxNB/derver/commit/0b76ccdfda9bdef2440b089706e860ce8a10ccce)
- update dapendencies [`21174900`](https://github.com/AlexxNB/derver/commit/21174900c30c804640dbd97bc2b07664374e50db)

## 0.4.15 - 2021-05-02

### Features

- Allow nested index fallbaks in SPA mode [`16e52070`](https://github.com/AlexxNB/derver/commit/16e52070e992cd252171099aeba12bcf0d419d7d)

    *If there is some sub directory contents index.html then unexistent URL with first part matched this subdirectory will fallback to this index file instead root one.*

### Bug Fixes

- 🐛 Last param from URL contents also a query part [`10c775ec`](https://github.com/AlexxNB/derver/commit/10c775ec9bd4968564d21212b9db77886ec4c11d)

## 0.4.14 - 2021-03-20

### Bug Fixes

- 🐛 Livereload doesn't work when no watchers, but uses remote. ([#1](https://github.com/AlexxNB/derver/issues/1)) [`dc79b251`](https://github.com/AlexxNB/derver/commit/dc79b251f1d518dc6c0a22a25cdae558f2e6ed84)

    *fix [#1](https://github.com/AlexxNB/derver/issues/1)*

## 0.4.13 - 2021-03-10

### Other

- fix 404 with query add usful data to request object [`226450e5`](https://github.com/AlexxNB/derver/commit/226450e5460c3db28951a2b38622a1f46aaf8c16)

## 0.4.12 - 2021-01-22

### Other

- dartheme for error modal [`a3e6214f`](https://github.com/AlexxNB/derver/commit/a3e6214fa548fa4ce23e3f26c42265a055605df3)

## 0.4.11 - 2021-01-21

### Other

- show modal when server down [`0c780542`](https://github.com/AlexxNB/derver/commit/0c7805429bf4e12cdcc9e5cdd7f01de14caea92e)

## 0.4.10 - 2021-01-21

### Other

- get remote config in each request [`fa781f69`](https://github.com/AlexxNB/derver/commit/fa781f69d81089f4b482f141e6668c16e4830ea5)

## 0.4.9 - 2021-01-21

### Other

- add remote server ID option [`809f22df`](https://github.com/AlexxNB/derver/commit/809f22dfd920e36767a8225fc7108b510abad760)

## 0.4.8 - 2021-01-21

### Other

- add remote control [`a815b00b`](https://github.com/AlexxNB/derver/commit/a815b00be5c735d3944047719fa3eeacafa0d2be)

## 0.4.7 - 2021-01-16

### Other

- fix nesting [`9a79fc26`](https://github.com/AlexxNB/derver/commit/9a79fc2644f868f6a7b7a6b7ea06d8613e356500)

## 0.4.6 - 2021-01-16

### Other

- add nested middlewares [`ab208b25`](https://github.com/AlexxNB/derver/commit/ab208b256f0f3fac59228ce03f30516af731e9bf)

## 0.4.5 - 2021-01-13

### Other

- fix middlewares wrong order [`fcbea885`](https://github.com/AlexxNB/derver/commit/fcbea885d696956f1c88cbd37d1e8f3c70374733)

## 0.4.4 - 2021-01-13

### Other

- fix chaining [`7b16b249`](https://github.com/AlexxNB/derver/commit/7b16b2498e469bab07c382910e773735c3cd6bd9)

## 0.4.3 - 2021-01-12

### Other

- debounce watch log [`b79943de`](https://github.com/AlexxNB/derver/commit/b79943deb4939be8ac2c6bdac7c3b686f48db6bc)

## 0.4.2 - 2021-01-12

### Other

- close watchers on exit [`701aa4e9`](https://github.com/AlexxNB/derver/commit/701aa4e9623a1ba43c4ba9d74a1bebadf91c2afa)

## 0.4.1 - 2021-01-12

### Other

- stop server on process exit [`91802948`](https://github.com/AlexxNB/derver/commit/9180294806ff3e8c352c403ab005aa6cc3676471)

## 0.4.0 - 2021-01-12

### Other

- add middlewares [`d80449e6`](https://github.com/AlexxNB/derver/commit/d80449e6387ae79f4e648e0b8015de859edbaad0)

## 0.3.0 - 2020-11-18

### Other


src/lib/livereload.js
import http from 'http';
import os from 'os';
import path from 'path';
import fs from 'fs';

const LR_URL = '/derver-livereload-events';
const LR_REMOTE_URL = '/derver-livereload-remote';

const listeners = new Set();

export function livereload(event,p1,p2){
  listeners.forEach(listener=>{
      if(typeof listener[event] == 'function') listener[event](p1,p2);
    });
}

export function createRemote(options){

    const remoteID = typeof options == 'string' ? options : false;

    let host = 'localhost';
    let port = 7000;

    if(!remoteID){
        options && options.host && (host = options.host)
        options && options.port && (port = options.port)
    }

    function sendCommand(command,data){
        return new Promise((resolve)=>{

            const config = remoteID && getRemoteConfig(remoteID);
            config && config.host && (hostname = config.host);
            config && config.port && (port = config.port)

            const req_options = {
                hostname: (config && config.host) || host,
                port: (config && config.port) || port,
                path: LR_REMOTE_URL,
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                }
            }
            const req = http.request(req_options, (res)=>{
                res.on('data',(chunk)=>{
                    if(chunk.toString()==='REMOTE OK')
                        resolve('OK')
                    else {
                        console.log('[Derver remote]: Warning: wrong command ' + command)
                        resolve('WARNING');
                    }
                });
            });
            req.on('error', (e)=>{
              console.log('[Derver remote]: Warning:' + e.message)
                resolve('WARNING');
            });
            req.write(JSON.stringify({command,data:data||{}}));
            req.end();
        });
    }

    return {
        reload(){return sendCommand('reload')},
        console(text){return sendCommand('console',{text})},
        error(text,header){return sendCommand('error',{text,header})}
    }
}

export function mwLivereload(options){
    if(!options.watch && !options.remote) return null;
    return function(req,res,next){
        if(req.url == LR_URL){

            const write = (evnt,data)=>res.write(`event: ${evnt}\ndata: ${JSON.stringify(data||{})}\n\n`)

            const listener = {
                reload: ()=>write('refresh'),
                console: (text)=>write('console',{text}),
                error: (text,header)=>write('srverror',{text,header:(header||'Error')}),
            }

            listeners.add(listener);

            res.writeHead(200, {
                'Content-Type': 'text/event-stream',
                'Cache-Control': 'no-cache',
                'Connection': 'keep-alive',
            });

            res.on('close', function() {
              listeners.delete(listener);
            });
            res.write('data: connected\n\n')
        }else if(options.remote && req.url == LR_REMOTE_URL){
            if(req.method == 'POST'){
                let json = '';

                req.on('data', chunk => {
                    json += chunk.toString();
                });

                req.on('end', () => {
                    const request = JSON.parse(json||'{}');

                    if(request.command == 'reload') livereload('reload');
                    else if(request.command == 'console') livereload('console',request.data.text);
                    else if(request.command == 'error') livereload('error',request.data.text,request.data.header);
                    else return res.end('REMOTE WRONG COMMAND');

                    res.end('REMOTE OK');
                });
            }else next();
        } else next();
    }
}

export function getLrURL(){
    return LR_URL;
}

export function getRemoteURL(){
    return LR_REMOTE_URL;
}

// Must be a clean function. Will be injected in browser client in <script> tag.
function lrClient(options){
    return (function(URL,scrollTimeout){
        let timer;

        function livereload(){
          if (!!window.EventSource) {
            var source = new EventSource(URL);

            function getData(e){
              return JSON.parse(e.data);
            }
        
            source.addEventListener('refresh', function(e) {
                window.location.reload(true);
            }, false)

            source.addEventListener('console', function(e) {
                console.log(getData(e).text);
            }, false)

            source.addEventListener('srverror', function(e) {
                let data = getData(e);
                showModal(data.header,data.text);
            }, false);
        
            source.addEventListener('open', function(e) {
              if(timer) location.reload();
              console.log('[Livereload] Ready')
            }, false)

            source.addEventListener('error', function(e) {

              if (e.eventPhase == EventSource.CLOSED) source.close();

              
              if (e.target.readyState == EventSource.CLOSED) {
                console.log("[Livereload] Disconnected! Retry in 5s...");
                !timer && showModal('Disconnected!','Connection with server was lost.');
                timer = setTimeout(livereload,5000);
              }else if (e.target.readyState == EventSource.CONNECTING) {
                console.log("[Livereload] Connecting...");
              }
            }, false);
          } else {
            console.error("[Livereload] Can't start Livereload! Your browser doesn't support SSE")
          }
        }

        function showModal(header,text){
          const message = document.createElement('div');
                message.innerHTML = `
                  <div class="lrmsg-bg">
                    <div class="lrmsg-modal">
                      <div class="lrmsg-close" onclick="this.parentNode.parentNode.remove()">×</div>
                      <div class="lrmsg-header">${header}</div>
                      <div class="lrmsg-content">${text}</div>
                    </div>
                  </div>
                  <style>
                    .lrmsg-bg{
                      font-family: Verdana, Geneva, sans-serif;
                      font-size: 16px;
                      background: rgba(30, 30, 30, 0.6);
                      position: fixed;
                      top: 0;
                      right: 0;
                      bottom: 0;
                      left: 0;
                      z-index: 1;
                    }

                    .lrmsg-modal{
                      position: relative;
                      max-width: 600px;
                      max-height: 400px;
                      margin: 40px auto; 
                      margin-top: 0px;
                      background-color: #1e1e1e;
                      border-top: 3px solid red;
                      border-radius: 5px;
                      opacity: 0;
                      animation: slide 0.3s forwards;
                      color: #cccccc;
                    }

                    .lrmsg-header{
                      font-weight: bold;
                      font-size: 18px;
                      padding: 10px;
                    }

                    .lrmsg-close{
                      float: right;
                      font-weight: bold;
                      color: #cccccc;
                      font-size: 25px;
                      margin: 3px 10px;
                      cursor: pointer;
                    }

                    .lrmsg-close:hover{color:#9a9a9a}

                    .lrmsg-content{
                      padding: 10px;
                      border-top: 1px solid #363636;
                    }

                    @keyframes slide {
                      100% { margin-top: 40px; opacity: 1;}
                  }
                  </style>
                `;

                document.body.append(message);
        }

        function preserveScroll(){
            const x = sessionStorage.getItem("_ds_x") || 0;
            const y = sessionStorage.getItem("_ds_y") || 0;
            setTimeout( ()=>window.scrollTo(x, y) ,scrollTimeout);

            window.addEventListener("scroll", ()=>{
              sessionStorage.setItem("_ds_x", window.scrollX);
              sessionStorage.setItem("_ds_y", window.scrollY);
            });
        }

        livereload();
        scrollTimeout && preserveScroll();

    }).toString(); 
};

export function mwInjectLivereload(options){
  
  if(!options.watch && !options.remote) return null;
  return function(req,res,next){
    if(['.html','.htm'].includes(req.extname)){
      res.body = Buffer.from(
        res.body.toString('utf-8').replace(
            /(<\/body>)/,
            `<script>(${lrClient(options)})('${LR_URL}',${options.preserveScroll === true ? 10 : Number(options.preserveScroll)})</script>\n$1`
        )
      );
    }
    
    next();
  }
}

function getRemoteConfig(name){
    const tmp = os.tmpdir();
    const file = path.join(tmp,'derver_'+name);
    if(!fs.existsSync(file)) return false;
    return JSON.parse(fs.readFileSync(file,'utf-8'));
}